core/common/src/main/java/alluxio/grpc/GrpcManagedChannelPool.java

```diff
@@ -63,120 +62,107 @@ public static GrpcManagedChannelPool INSTANCE() {
-        ManagedChannelReference channelReference = channelEntry.getValue();
-        if (channelReference.getRefCount() <= 0) {
-          mChannels.remove(channelKey);
-          channelsToDestroy.add(new Pair(channelKey, channelReference));
-        }
-      }
-    }
-
-    // TODO(ggezer) Consider shutting down in parallel.
-    for (Pair<ChannelKey, ManagedChannelReference> channelPair : channelsToDestroy) {
-      ManagedChannel channel = channelPair.getSecond().get();
-      channel.shutdown();
+  private boolean waitForChannelReady(ManagedChannel managedChannel) {
+    ConnectivityState initialState = managedChannel.getState(false);
+    if (initialState != ConnectivityState.READY) {
       try {
-        channel.awaitTermination(
-            Configuration.getMs(PropertyKey.MASTER_GRPC_CHANNEL_SHUTDOWN_TIMEOUT),
-            TimeUnit.MILLISECONDS);
+        CommonUtils.waitFor(this + " to start",
+            () -> managedChannel.getState(true) == ConnectivityState.READY,
+            WaitForOptions.defaults().setTimeoutMs((int) mHealthCheckTimeoutMs));
       } catch (InterruptedException e) {
         Thread.currentThread().interrupt();
-        // Allow thread to exit.
-      } finally {
-        channel.shutdownNow();
+        return false;
+      } catch (TimeoutException e) {
+        return false;
       }
-      Verify.verify(channel.isShutdown());
-      destroyedCount++;
-      LOG.debug("Destroyed gRPC managed channel for key:{}", channelPair.getFirst());
     }
-    LOG.debug("gRPC channel terminator shut down {} of {} total channels.", destroyedCount,
-        channelCount);
+    return true;
   }
 
   /**
-   * Acquires and Increases the reference count for the singleton {@link ManagedChannel}.
+   * Acquires and increases the ref-count for the {@link ManagedChannel}.
    *
    * @param channelKey channel key
    * @return a {@link ManagedChannel}
    */
   public ManagedChannel acquireManagedChannel(ChannelKey channelKey) {
+    boolean shutdownExistingChannel = false;
     try (LockResource lockShared = new LockResource(mLock.readLock())) {
       if (mChannels.containsKey(channelKey)) {
-        return mChannels.get(channelKey).reference();
+        ManagedChannelReference managedChannelRef = mChannels.get(channelKey);
+        if (waitForChannelReady(managedChannelRef.get())) {
+          return managedChannelRef.reference();
+        } else {
+          // Postpone channel shutdown under exclusive lock below.
+          shutdownExistingChannel = true;
+        }
       }
     }
     try (LockResource lockExclusive = new LockResource(mLock.writeLock())) {
+      // Dispose existing channel if required.
+      int existingRefCount = 0;
+      if (shutdownExistingChannel && mChannels.containsKey(channelKey)) {
+        shutdownManagedChannel(mChannels.get(channelKey).get());
+        existingRefCount = mChannels.get(channelKey).getRefCount();
+        mChannels.remove(channelKey);
+      }
```

Comment:
We could probably do the shutdown portion outside the lock and just remove in the critical section?

Reply:
Then we wouldn't be serializing shutting down of the channels!

Reply:
Let me try it though. We may not hit the same issue anymore...

Reply:
Scratch that. I think this will cause calling shutdown concurrently on the same channel.
I suggest putting a TODO and trying this at a later time.

Reply:
Sounds good